File: cpswAsynPortDriverApp/deviceLibrarySrc/PrbsTx.cc
#include "PrbsTx.h"

#define addrSize 4

PrbsTx IPrbsTx::create(const char *name)
{
PrbsTxImpl v = CShObj::create<PrbsTxImpl>(name);
Field f;
        f = IIntField::create("ConfigStatus", 32, false, 0);
        v->CMMIODevImpl::addAtAddress( f , 0x00*addrSize);
        f = IIntField::create("AxiEn", 1, false, 0);
        v->CMMIODevImpl::addAtAddress( f , 0x00*addrSize);
        f = IIntField::create("TxEn", 1, false, 1);
        v->CMMIODevImpl::addAtAddress( f , 0x00*addrSize);
        f = IIntField::create("Busy", 1, false, 2);
        v->CMMIODevImpl::addAtAddress( f , 0x00*addrSize);
        f = IIntField::create("Overflow", 1, false, 3);
        v->CMMIODevImpl::addAtAddress( f , 0x00*addrSize);
        f = IIntField::create("PacketLength", 32, false, 0);
        v->CMMIODevImpl::addAtAddress( f , 0x01*addrSize);
        f = IIntField::create("AxiStreamCfg", 32, false, 0);
        v->CMMIODevImpl::addAtAddress( f , 0x02*addrSize);
        f = IIntField::create("tDest", 8, false, 0);
        v->CMMIODevImpl::addAtAddress( f , 0x02*addrSize);
        f = IIntField::create("tId", 8, false, 8);
        v->CMMIODevImpl::addAtAddress( f , 0x02*addrSize);
        f = IIntField::create("DataCount", 32, false, 0);
        v->CMMIODevImpl::addAtAddress( f , 0x03*addrSize);
        f = IIntField::create("EventCount", 32, false, 0);
        v->CMMIODevImpl::addAtAddress( f , 0x04*addrSize);
        f = IIntField::create("RandomData", 32, false, 0);
        v->CMMIODevImpl::addAtAddress( f , 0x05*addrSize);
        return v;
}

CPrbsTxImpl::CPrbsTxImpl(Key &key, const char *name) : CMMIODevImpl(key, name, 0x00010000 >> 2, LE)
{
}

File: util/templates/devices/PrbsTx.py
from __future__ import print_function
import os
from Cheetah.Template import Template

d = { 'name': "PrbsTx",
     'registers': [{ 'name'   : "ConfigStatus",
                    'address' : "0x00",
                    'type'    : "Configuration" },
                   { 'name'   : "AxiEn",
                     'address': "0x00",
                     'size'   : "1",
                     'lsBit'  : "0",
                     'type'   : "Configuration"},
                   { 'name'   : "TxEn",
                     'address': "0x00",
                     'size'   : "1",
                     'lsBit'  : "1",
                     'type'   : "Configuration"},
                   { 'name'   : "Busy",
                     'address': "0x00",
                     'size'   : "1",
                     'lsBit'  : "2",
                     'type'   : "Status"},
                   { 'name'   : "Overflow",
                     'address': "0x00",
                     'size'   : "1",
                     'lsBit'  : "3",
                     'type'   : "Status"},
                   {'name'    : "PacketLength",
                    'address' : "0x01",
                    'type'    : "Configuration" },
                   {'name'    : "AxiStreamCfg",
                    'address' : "0x02",
                    'type'    : "Configuration" },
                   { 'name'   : "tDest",
                     'address': "0x02",
                     'size'   : "8",
                     'lsBit'  : "0",
                     'type'   : "Configuration"},
                   { 'name'   : "tId",
                     'address': "0x02",
                     'size'   : "8",
                     'lsBit'  : "8",
                     'type'   : "Configuration"},
                   {'name'    : "DataCount",
                    'address' : "0x03",
                    'type'    : "Status" },
                   {'name'    : "EventCount",
                    'address' : "0x04",
                    'type'    : "Status" },
                   {'name'    : "RandomData",
                    'address' : "0x05",
                    'type'    : "Status" } ],
     'commands': [{ 'name'   : "OneShot"} ]
   }



os.mkdir('PrbsTx')
os.mkdir('PrbsTx/deviceLibrary')
os.mkdir('PrbsTx/PrbsTxSrc')
os.mkdir('PrbsTx/Db')
os.mkdir('PrbsTx/ui')

#device library
f=open('PrbsTx/deviceLibrary/PrbsTx.cc', 'w')
print(Template( file='../deviceTemplate/builder.cc.tmpl', searchList = [d] ), file=f)
f.close()

f=open('PrbsTx/deviceLibrary/PrbsTx.h', 'w')
print(Template( file='../deviceTemplate/builder.h.tmpl', searchList = [d] ), file=f)
f.close()

#asyn driver
f=open('PrbsTx/PrbsTxSrc/Makefile', 'w')
print(Template( file='../deviceTemplate/Makefile.tmpl', searchList = [d] ), file=f)
f.close()

f=open('PrbsTx/PrbsTxSrc/PrbsTxDriver.cpp', 'w')
print(Template( file='../deviceTemplate/templateDriver.cpp.tmpl', searchList = [d] ), file=f)
f.close()

f=open('PrbsTx/PrbsTxSrc/PrbsTxDriver.h', 'w')
print(Template( file='../deviceTemplate/templateDriver.h.tmpl', searchList = [d] ), file=f)
f.close()

f=open('PrbsTx/PrbsTxSrc/PrbsTxDriverSupportInclude.dbd', 'w')
print(Template( file='../deviceTemplate/templateDriverSupportInclude.dbd.tmpl', searchList = [d] ), file=f)
f.close()

#database substitutions
f=open('PrbsTx/Db/PrbsTx.substitutions', 'w')
print(Template( file='../deviceTemplate/template.substitutions.tmpl', searchList = [d] ), file=f)
f.close()

#gui
f=open('PrbsTx/ui/PrbsTx.ui', 'w')
print(Template( file='../deviceTemplate/templateGui.ui.tmpl', searchList = [d] ), file=f)
f.close()
